### DISCUSSION TITLE
Is there a minimal working example to use smeagol-galore with a LDAP server for user auth?

### DISCUSSION THREAD

**Discussion by Tony2** [schnatterer/smeagol-galore#10] [Q&A]
Hi,

I read that LDAP auth is possible via CAS. Is is possible to have a minimal working example how to run smeagol-galore with LDAP auth (for example with JumpCloud)? It would be much appreciated, since I have never used CAS before.

**Comment by schnatterer** [comment]
Hi @Tony2,

 [cloudogu/cas](https://github.com/cloudogu/cas/) uses LDAP. It's not minimal, and uses openldap (not jumpcloud), though.
Note that they're updating to cas 6 at moment, so I presume smeagol galore will also upgrade in the not so far future.
For now, smeagol galore uses cas 4, so maybe [the docs](https://apereo.github.io/cas/4.0.x/index.html) can help.
Closing this because I consider it more of a question than an actual issue.

**Comment by Tony2** [comment]
Hi @schnatterer,

thank you for the hint. Here is my understanding of what is needed to be done to get LDAP auth:

- clone this repo
- make change to deployerConfigContext.xml, as instructed in https://apereo.github.io/cas/4.1.x/installation/LDAP-Authentication.html
- build the docker image and start a container from it, instead of the official docker image

Is this correct, or am I still missing some step?

**Reply by schnatterer** [reply]
Well in general checking out and building your own image shouldn't be necessary.
You could just copy the deployerConfigContext.xml from the repo (or image, or container) adapt it to your needs and bind mount it in a container started from the official image to /tomcat/webapps/cas/WEB-INF/deployerConfigContext.xml.

BUT: Your link to the cas docs mention a special dependency that is not included in smeagol right now. 
So for now, you're right, you'll have to clone, add the dependency to cas/pom.xml and build the image. Then do the bind mounting on container start.
If this works for you, I'll gladly accept a PR, so you won't need your own image in the long run. A simple reproducible example (e.g. with an openldap container) would be nice, so I can test this before integrating.

**Reply by hth2** [reply]
I did the following:

- add dependency for LDAP to  cas/pom.xml
- add relevant config to  cas/src/main/webapp/WEB-INF/deployerConfigContext.xml
- create  a file cas/etc/cas/ldap.properties as suggested by CAS docs
-  add cas/etc/cas/ZeroSSL-CA.cer (the ldap server uses Let's Encrypt cert, this is the relevant CA)
- modify cas/src/main/webapp/WEB-INF/spring-configuration/propertyFileConfigurer.xml to load ldap.properties

After that I could successfully login with a credential from LDAP
`2021-07-18 14:51:54,777 INFO [org.jasig.cas.authentication.PolicyBasedAuthenticationManager] - <Authenticated testuser with credentials [testuser+password].>`

But after login I got some error:
![image](https://user-images.githubusercontent.com/1019040/126072223-3a6294ed-4f75-4188-989b-2c56d78a3fed.png)

and the log says:
```
org.apache.shiro.authc.AuthenticationException: Authentication token of type [class org.apache.shiro.authc.UsernamePasswordToken] could not be authenticated by any configured realms.  Please ensure that at least one realm can authenticate these tokens.
.
.
2021-07-18 14:51:57.089 [https-openssl-apr-8443-exec-10] ERROR org.springframework.boot.web.servlet.support.ErrorPageFilter - Forwarding to error page from request [/api/v1/repositories] due to exception [org.springframework.web.client.RestClientException: Could not extract response: no suitable HttpMessageConverter found for response type [class com.cloudogu.smeagol.repository.infrastructure.ScmRepositoryRepository$SCMRootEndpointDTO] and content type [text/html;charset=UTF-8]]
com.google.common.util.concurrent.UncheckedExecutionException: org.springframework.web.client.RestClientException: Could not extract response: no suitable HttpMessageConverter found for response type [class com.cloudogu.smeagol.repository.infrastructure.ScmRepositoryRepository$SCMRootEndpointDTO] and content type [text/html;charset=UTF-8]
.
.
Caused by: org.springframework.web.client.RestClientException: Could not extract response: no suitable HttpMessageConverter found for response type [class com.cloudogu.smeagol.repository.infrastructure.ScmRepositoryRepository$SCMRootEndpointDTO] and content type [text/html;charset=UTF-8]

```

I have no previous experience with CAS/Spring/cloudogu and kind of got stuck here. Any idea what I can try next?

**Reply by schnatterer** [reply]
Nice progress!
So you can log in successfully on https://localhost:8443/cas.
On https://localhost:8443/smeagol you encounter the error stated above.
Correct?
Do you also get an error on https://localhost:8443/scm?
My experience is that the smeagol -> SCMM communictation often causes a lot of errors and it's easier to get SCMM working properly first.
Did you check if [cloudogu/cas](https://github.com/cloudogu/cas/blob/develop/app/pom.xml) adds more dependencies?
